client.go

```diff
@@ -273,30 +272,75 @@ func (c *Client) DoUpdate(ctx context.Context, creds Credentials) ([]byte, []byt
 	return result.Config, dhPrivkeyPEM, newCreds, nil
 }
 
-// postDNClient wraps and signs the given dnclientRequestWrapper message, and makes the API call.
-// On success, it returns the response message body. On error, the error is returned.
-func (c *Client) postDNClient(ctx context.Context, reqType string, value []byte, hostID string, counter uint, privkey ed25519.PrivateKey) ([]byte, error) {
-	encMsg, err := json.Marshal(message.RequestWrapper{
-		Type:      reqType,
-		Value:     value,
-		Timestamp: time.Now(),
+func (c *Client) StreamLogs(ctx context.Context, creds Credentials, responseToken string) (*StreamController, error) {
+	value, err := json.Marshal(message.CommandResponseRequest{
+		ResponseToken: responseToken,
 	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal DNClient message: %s", err)
+	}
+
+	return c.streamingPostDNClient(ctx, message.CommandResponse, value, creds.HostID, creds.Counter, creds.PrivateKey)
+}
+
+func (c *Client) streamingPostDNClient(ctx context.Context, reqType string, value []byte, hostID string, counter uint, privkey ed25519.PrivateKey) (*StreamController, error) {
+	pr, pw := io.Pipe()
+
+	postBody, err := SignRequestV1(reqType, value, hostID, counter, privkey)
 	if err != nil {
 		return nil, err
 	}
-	signedMsg := base64.StdEncoding.EncodeToString(encMsg)
-	sig := ed25519.Sign(privkey, []byte(signedMsg))
-	body := message.RequestV1{
-		Version:   1,
-		HostID:    hostID,
-		Counter:   counter,
-		Message:   signedMsg,
-		Signature: sig,
+	pbb := bytes.NewBuffer(postBody)
+
+	req, err := http.NewRequestWithContext(ctx, "POST", c.dnServer+message.EndpointV1, io.MultiReader(pbb, pr))
+	if err != nil {
+		return nil, err
 	}
-	postBody, err := json.Marshal(body)
+
+	done := make(chan struct{})
+	sc := &StreamController{w: pw, done: done}
+
+	go func() {
+		resp, err := c.http.Do(req)
```

Comment:
this http client is configured with a 1 minute timeout

Reply:
![image](https://github.com/DefinedNet/dnapi/assets/440033/b08438b8-43ca-4a6e-a158-ef0b77aa9148)

I think for streaming requests, we want a very long `http.Client.Timeout` - or none at all. Do we want the server to control how long logs stream for via an arg? If so, this timeout will always be a maximum boundary.

By default, no timeouts are set, so if we remove or increase `http.Client.Timeout`, we'll need to set other timeouts appropriately.

`http.Transport.IdleConnTimeout` is about how long a connection stays in the connection pool, so we can safely ignore this one. The rest however...

- In our backend services, we use a `net.Dialer.Timeout` of 500ms - but this is probably too slow for some network conditions, e.g. when on mobile. Perhaps 5s makes sense? Or call it 30s? What's the worst that happens?
- `http.Transport.TLSHandshakeTimeout` 10s?
- `http.Transport.ResponseHeaderTimeout` 10s?
- `http.Client.Timeout` 15 minutes?

We could use a second HTTP client with these timeouts just for streaming - or we could adopt these timeouts for the non-streaming calls as well. If we want to make sure non-streaming calls return faster than 15 minutes we can make use of context deadlines.

Reply:
dnapi has some XWithTimeout functions that I kind of regret (I think it would've made more sense to set the deadline at the caller rather than having a utility function inside the package - and I might update mobile to do this, since it's the only caller.)

```
❯ cd ../mobile_nebula
❯ rg WithTimeout
nebula/api.go
51:43:	cfg, pkey, creds, meta, err := c.c.EnrollWithTimeout(context.Background(), 30*time.Second, c.l, code)
102:44:	updateAvailable, err := c.c.CheckForUpdateWithTimeout(context.Background(), 10*time.Second, creds)
115:42:	cfg, pkey, newCreds, err := c.c.DoUpdateWithTimeout(context.Background(), 10*time.Second, creds)
```

What do you think about a long (15 minute) timeout on the client, and updating dnclient to manually set shorter, applicable timeouts for these calls?

Reply:
Could keep the Transport on dnapi.Client, and then the streaming responder can create a new http.Client either with a command-specific timeout, or a manually controlled timeout (as described in the article - something that closes requests that get stuck wherever.)

```
// Client communicates with the API server.
type Client struct {
	http     *http.Client
	transport *uaTransport
	dnServer string
}

// NewClient returns new Client configured with the given useragent.
// It also supports reading Proxy information from the environment.
func NewClient(useragent string, dnServer string) *Client {
	transport: &uaTransport{
				T: &http.Transport{
					Proxy: http.ProxyFromEnvironment,
				},
				useragent: useragent,
			}
	return &Client{
		transport: transport,
		http: &http.Client{
			Timeout: 1 * time.Minute,
			Transport: transport,
		},
		dnServer: dnServer,
	}
}
```

Reply:
Oops, just saw this. I pushed a commit that changes the overall request timeout to 15 minutes, and sets 10s, 10s, and 10s for Dial, TLS handshake, and response headers. My next step was going to be to update dnclient and mobile to set their own context deadlines appropriately (I already did this in tests & the example.) I liked the idea of pushing this back on the caller, since our two callers have different needs.

> or a manually controlled timeout (as described in the article - something that closes requests that get stuck wherever.)
I think the context deadline is the way to do this - but I didn't think a deadline of 15 minutes passed to `Client.Do` would override the `Client.Timeout` of 1 minute already set. That was the motivation for changing the overall timeout.

If that's correct, I think the only real alternative to this approach is to change the Client.Timeout back to 1 minute and use a second Client for streaming. Is that what you prefer? I'm open to whatever.

Reply:
You can see the timeouts mobile uses here: https://github.com/DefinedNet/dnapi/pull/9#discussion_r1608894026

I think Enroll has a shorter timeout for UI/UX benefits (not sitting waiting a full minute.) The others being 10 seconds, IDK. We can bump them all up to 60 seconds, and take the minor UX hit.

Reply:
I think I've updated it with what you'd like to see, let me know if not. I know have two HTTP clients, one for Enroll & standard dnclient calls, and a second for streaming dnclient calls. I left the 10s Dial, TLS, and response header timeouts in place for both clients. The overall request timeout is 60 seconds for the former and 15 minutes for the latter. Clients can still use context to set these timeouts lower if they wish.